juniper_official/Linecard/check-pfe-discards.rule

```diff
@@ -0,0 +1,321 @@
+/*
+ * Detects PFE discards and notifies when anomalies are found.
+ * One inputs control detection
+ * 
+ *   1) discard-count, is the threshold that causes the rule to report
+ *      an anomaly.  By default it's 1. This rule will set a dashboard
+ *      color to red when *all* the discards are greater than
+ *      'discard-count' for 60s. otherwise color is set to green.
+ */  
+healthbot {
+    topic linecard.pfe {
+        rule check-pfe-discards {
+             /*
+             * Monitors the pfe discards count on PFE. Notifies via the dashboard
+             * colors when the flaps increase by {{discard-count}} for every collected
+             * metric during a 60s time range (red). Otherwise the color is set to green
+             * 
+             * Use FPC and PFE names as key for rule.
+             */  
+            keys [ fpc-no pfe-no ];
+            synopsis "Packet forwarding engine hardware discard statistics analyzer";
+            description "Collects packet forwarding engine hardware discard statistics  and notifies when discard count increases";
+            sensor pfe-sensor {
+                synopsis "iAgent netconf sensor definition";
+                description "Collects data from network device netconf rpc using get-pfe-traffic-statistics";
+                iAgent {
+                    file pfe-discard-statistics.yml;
+                    table PfeDiscardStatisticsTable;
+                    frequency 60s;
+                }
+            }
+            field drop-threshold {
+                constant {
+                    value "{{discard-count}}";
+                }
+                type integer;
+                description "Drop count increase threshold";
+            }
+            field fpc-no {
+                sensor pfe-sensor {
+                    path pfe-fpc-index;
+                }
+                type string;
+                description "FPC slot number";
+            }
+            field pfe-no {
+                sensor pfe-sensor {
+                    path pfe-index;
+                }
+                type string;
+                description "PFE number";
+            }
+            trigger data-error-discard {
+                synopsis "PFE data error discards KPI";
+                description "Sets health based on the change in PFE data error discard count";
+                frequency 60s;
+                term is-data-error-discard {
+                    when {
+                        increasing-at-least-by-value "$data-error-discard" {
```

Comment:
Would it make sense to declare all the iAgent YAML implicit fields in the rule explicitly to ensure that these can be plotted as part of the field table?

Reply:
It has been handled in iAgent hence we can plot graphs for all fields below;    fabric-discard: {fabric-discard: int}
    bits-to-test-discard: {bits-to-test-discard: int}
    data-error-discard: {data-error-discard: int}
    tcp-header-error-discard: {tcp-header-error-discard: int}
    stack-underflow-discard: {stack-underflow-discard: int}
    stack-overflow-discard: {stack-overflow-discard: int}
    bad-route-discard: {bad-route-discard: int}
    nexthop-discard: {nexthop-discard: int}
    invalid-iif-discard: {invalid-iif-discard: int}
    info-cell-discard: {info-cell-discard: int}